src/groups/GroupList.tsx

```diff
@@ -0,0 +1,146 @@
+import _ from 'lodash/fp';
+import React, { useEffect, useState } from 'react';
+import { DeleteConfirmationModal } from 'src/components/common';
+import FooterWrapper from 'src/components/FooterWrapper';
+import { Sort } from 'src/components/group-common';
+import { DelayedSearchInput } from 'src/components/input';
+import LeaveResourceModal from 'src/components/LeaveResourceModal';
+import { PageBox, PageBoxVariants } from 'src/components/PageBox';
+import { TopBar } from 'src/components/TopBar';
+import { GroupCard } from 'src/groups/GroupCard';
+import { GroupCardHeaders } from 'src/groups/GroupCardHeaders';
+import { NewGroupCard } from 'src/groups/NewGroupCard';
+import { NewGroupModal } from 'src/groups/NewGroupModal';
+import { NoGroupsMessage } from 'src/groups/NoGroupsMessage';
+import { Ajax } from 'src/libs/ajax';
+import { CurrentUserGroupMembership } from 'src/libs/ajax/Groups';
+import { reportError } from 'src/libs/error';
+import { useCancellation } from 'src/libs/react-utils';
+import * as StateHistory from 'src/libs/state-history';
+import * as Style from 'src/libs/style';
+import { cond, textMatch } from 'src/libs/utils';
+
+export const GroupList = (props: {}): React.ReactNode => {
+  const signal = useCancellation();
+
+  const fetchGroups = async (): Promise<CurrentUserGroupMembership[]> => {
+    const rawGroups = await Ajax(signal).Groups.list();
+    const updatedGroups = _.flow(
+      _.groupBy('groupName'),
+      _.map((gs) => ({ ...gs[0], role: _.map('role', gs) })),
+      _.sortBy('groupName')
+    )(rawGroups);
+    return updatedGroups;
+  };
+
+  // State
+  const [filter, setFilter] = useState<string>(() => StateHistory.get()?.filter || '');
+  const [creatingNewGroup, setCreatingNewGroup] = useState(false);
+  const [deletingGroup, setDeletingGroup] = useState<CurrentUserGroupMembership>();
+  const [leavingGroup, setLeavingGroup] = useState<CurrentUserGroupMembership>();
+  const [sort, setSort] = useState<Sort>({ field: 'groupName', direction: 'asc' });
+  const [groups, updateGroups] = useAutoLoadedData<CurrentUserGroupMembership[]>(fetchGroups, [signal], {
+    onError: (state) => reportError('Error loading group list', state.error),
+  });
+
+  // Helpers
+  const refresh = async () => {
+    setCreatingNewGroup(false);
+    setDeletingGroup(undefined);
+    setLeavingGroup(undefined);
+    updateGroups(fetchGroups);
+  };
+
+  const doDelete = (groupName: string) =>
+    updateGroups(async () => {
+      setDeletingGroup(undefined);
+      try {
+        await Ajax().Groups.group(groupName).delete();
+      } catch (e) {
+        // this is in a separate try/catch on its own, so that a failure in deleting won't error the entire page
+        reportError('Error deleting group.', e);
+      }
+      const updatedGroups = await fetchGroups();
+      return updatedGroups;
+    });
+
+  useEffect(() => {
+    StateHistory.update({ filter });
```

Comment:
This was also saving the groups in state history, as well as initializing the group state with it. But this seems very unnecessary since the groups are always fetched when the component is first mounted. It's also potentially a lot to store, not to mention the potential cause of buggy behavior if there are object format changes, api call failures, etc.